File: python/factory.py
# This example demonstrates factory design pattern in Python.

class Person:
    def __init__(self, name):
        self.__name = name
    
    def get_name(self):
        return self.__name
    
    def set_name(self, name):
        self.__name = name

class Employee(Person):
    def __init__(self, name, salary):
        super().__init__(name)
        self.__salary = salary
    
    def get_name(self):
        return super().get_name() + " (Employee)"
    
    def get_salary(self):
        return self.__salary
    
    def set_salary(self, salary):
        self.__salary = salary

class Customer(Person):
    def __init__(self, name, email):
        super().__init__(name)
        self.__email = email
    
    def get_name(self):
        return super().get_name() + " (Customer)"

    def get_email(self):
        return self.__email
    
    def set_email(self, email):
        self.__email = email

class PersonFactory:
    @staticmethod
    def create_person(type):
        if type == "employee":
            return Employee("John", 10000)
        elif type == "customer":
            return Customer("Jack", "abc@somedomain.com")
        else:
            return Person("")
        
p1 = PersonFactory.create_person("employee")
print(p1.get_name())

p2 = PersonFactory.create_person("customer")
print(p2.get_name())

p3 = PersonFactory.create_person("guest")
print(p3.get_name())

# Output:
# John (Employee)
# Jack (Customer)
#   

# The PersonFactory class is a factory class that creates objects of the Person class and its subclasses.
# The create_person() method of the PersonFactory class creates and returns objects of the Person class and its subclasses.
# The create_person() method takes a type parameter and returns an object of the Person class or its subclasses based on the type parameter.
# The create_person() method returns an object of the Person class if the type parameter is not "employee" or "customer".
# The create_person() method returns an object of the Employee class if the type parameter is "employee".
# The create_person() method returns an object of the Customer class if the type parameter is "customer".

# The factory design pattern is used to create objects of a class without exposing the creation logic to the client.
# The factory design pattern is used when we have a super class with multiple sub-classes and based on input, we need to return one of the sub-class.

# This example demonstrates factory design pattern in Python.

File: javascript/factory.js
// This example demostrate the factory in javascript
// The factory is a design pattern that creates objects without exposing the instantiation logic to the client.
// The factory design pattern is useful when we have a super class with multiple sub-classes and based on input, we need to return one of the sub-class.
// The factory design pattern provides approach to code for interface rather than implementation.

// Example 1
// The following example demostrate the factory
function Person(name, age) {
    this.name = name;
    this.age = age;
}

Person.prototype.getName = function() {
    return this.name;
}

function Employee(name, age, salary) {
    Person.call(this, name, age);
    this.salary = salary;
}

Employee.prototype = Object.create(Person.prototype);
Employee.prototype.constructor = Employee;

Employee.prototype.getSalary = function() {
    return this.salary;
}

function Customer(name, age, address) {
    Person.call(this, name, age);
    this.address = address;
}

Customer.prototype = Object.create(Person.prototype);
Customer.prototype.constructor = Customer;

Customer.prototype.getAddress = function() {
    return this.address;
}

function Factory() {
    this.create = function(type, name, age, value) {
        switch (type) {
            case "employee":
                return new Employee(name, age, value);
                break;
            case "customer":
                return new Customer(name, age, value);
                break;
            default:
                return new Person(name, age);
        }
    }
}

var factory = new Factory();
var employee = factory.create("employee", "John", 30, 10000);
console.log(employee.getName());
console.log(employee.getSalary());

var customer = factory.create("customer", "John", 30, "New York");
console.log(customer.getName());
console.log(customer.getAddress());

// Output:
// John
// 10000
// John
// New York